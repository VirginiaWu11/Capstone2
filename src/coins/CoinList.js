import { useEffect, useState, memo, useCallback } from "react";
import CoinGeckoApi from "../CoinGeckoApi";
import CoinsCardList from "./CoinsCardList";
import Grid from "@mui/material/Grid";
import Container from "@mui/material/Container";
import Box from "@mui/material/Box";
import CoinTable from "./CoinTable";
import ToggleButton from "@mui/material/ToggleButton";
import ToggleButtonGroup from "@mui/material/ToggleButtonGroup";
import ViewListIcon from "@mui/icons-material/ViewList";
import ViewModuleIcon from "@mui/icons-material/ViewModule";
import InputLabel from "@mui/material/InputLabel";
import MenuItem from "@mui/material/MenuItem";
import FormControl from "@mui/material/FormControl";
import Select from "@mui/material/Select";
import Pagination from "@mui/material/Pagination";
import Stack from "@mui/material/Stack";

const CoinList = memo(({ watchlistIds, handleOpen }) => {
  const [isLoading, setIsLoading] = useState(true);
  const [coins, setCoins] = useState([]);
  const [coinsToggleView, setCoinsToggleView] = useState("allcoins");
  const [listModuleView, setlistModuleView] = useState("module");
  const [itemsPerPage, setItemsPerPage] = useState(16);
  const [page, setPage] = useState(1);
  console.debug("watchlistIds in CoinList", watchlistIds);

  const getCoins = useCallback(async (page, itemsPerPage) => {
    const resp = await CoinGeckoApi.getCoins(page, itemsPerPage);
    setCoins(resp);
  }, []);

  const getCoinInfoCoinGecko = useCallback(
    async (page, itemsPerPage, watchlistIds) => {
      const resp = await CoinGeckoApi.getCoins(1, 1, watchlistIds);
      let divisor = watchlistIds?.length / itemsPerPage;
      setCoins(
        resp.slice(
          (watchlistIds?.length / divisor) * (page - 1),
          (watchlistIds?.length / divisor) * page
        )
      );
    },
    []
  );

  const handleChange = useCallback(
    (event, nextView) => {
      setPage(1);
      if (nextView === "allcoins") {
        getCoins(1, itemsPerPage);
      } else {
        getCoinInfoCoinGecko(1, itemsPerPage, watchlistIds);
      }
      setCoinsToggleView(nextView);
    },
    [getCoinInfoCoinGecko, getCoins, itemsPerPage, watchlistIds, setPage]
  );

  useEffect(() => {
    if (coinsToggleView === "allcoins") {
      getCoins(page, itemsPerPage);
    } else {
      getCoinInfoCoinGecko(page, itemsPerPage, watchlistIds);
    }
    setIsLoading(false);
  }, [
    coinsToggleView,
    handleChange,
    page,
    getCoinInfoCoinGecko,
    getCoins,
    itemsPerPage,
    watchlistIds,
  ]);
  if (isLoading) {
    return <p>Loading &hellip;</p>;
  }

  const ListModuleToggleButtons = memo(() => {
    const handleChange = (event, nextView) => {
      setlistModuleView(nextView);
    };

    return (
      <ToggleButtonGroup
        value={listModuleView}
        exclusive
        onChange={handleChange}
        sx={{ ml: 2, mt: 2 }}
      >
        <ToggleButton value="list" aria-label="list">
          <ViewListIcon />
        </ToggleButton>
        <ToggleButton value="module" aria-label="module">
          <ViewModuleIcon />
        </ToggleButton>
      </ToggleButtonGroup>
    );
  });

  const NumberOfItemsSelect = () => {
    const handleChange = (event) => {
      setItemsPerPage(event.target.value);
    };

    return (
      <Box sx={{ minWidth: 150, mx: 2, mb: 1 }}>
        <FormControl fullWidth size="small">
          <InputLabel id="items-per-page">Items Per Page</InputLabel>
          <Select
            labelId="items-per-page"
            id="items-per-page"
            value={itemsPerPage}
            label="items per page"
            onChange={handleChange}
          >
            <MenuItem value={10}>Ten</MenuItem>
            <MenuItem value={16}>Sixteen</MenuItem>
            <MenuItem value={24}>TwentyFour</MenuItem>
          </Select>
        </FormControl>
      </Box>
    );
  };

  const PaginationOutlined = memo(({ itemsPerPage, totalItems = 0 }) => {
    const handleChange = useCallback((event, value) => {
      setPage(value);
    }, []);
    return (
      <Stack spacing={2}>
        <Pagination
          count={Math.ceil(totalItems / itemsPerPage)}
          page={page}
          onChange={handleChange}
          variant="outlined"
          color="primary"
        />
      </Stack>
    );
  });

  const FilterCoinsToggleButtons = memo(() => {
    return (
      <ToggleButtonGroup
        value={coinsToggleView}
        exclusive
        onChange={handleChange}
        sx={{ mr: 2, mt: 2 }}
      >
        <ToggleButton value="allcoins" aria-label="all coins">
          All Coins
        </ToggleButton>
        <ToggleButton value="watchlist" aria-label="watchllist">
          Watchlist
        </ToggleButton>
      </ToggleButtonGroup>
    );
  });

  return (
    <div>
      <Container maxWidth="xl">
        <Box sx={{ display: "flex", justifyContent: "space-between" }}>
          <ListModuleToggleButtons />
          <FilterCoinsToggleButtons />
        </Box>
        {listModuleView === "module" ? (
          coins.length ? (
            <CoinsCardList
              coins={coins}
              watchlistIds={watchlistIds}
              handleOpen={handleOpen}
            />
          ) : (
            <Grid item xs={3}>
              <p>No coins found</p>
            </Grid>
          )
        ) : (
          <CoinTable
            coins={coins}
            watchlistIds={watchlistIds}
            handleOpen={handleOpen}
          />
        )}
        <Grid container spacing={1} justifyContent="center" pt={2}>
          <NumberOfItemsSelect />
          <PaginationOutlined
            itemsPerPage={itemsPerPage}
            totalItems={
              coinsToggleView === "watchlist" ? watchlistIds?.length : 11995
            }
          />
        </Grid>
      </Container>
    </div>
  );
});

export default CoinList;
