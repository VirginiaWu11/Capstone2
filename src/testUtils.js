import React from "react";
import { UserContext } from "./auth/UserContext";

const demoUser = {
  username: "testuser",
  first_name: "testfirst",
  last_name: "testlast",
  email: "test@test.net",
};

const UserProvider = ({
  children,
  currentUser = demoUser,
  setCurrentUser,
  token,
  setToken,
  signup,
  signin,
  signout,
  updateProfile,
  infoLoaded,
  setInfoLoaded,
}) => (
  <UserContext.Provider
    value={{
      currentUser,
      setCurrentUser,
      token,
      setToken,
      signup,
      signin,
      signout,
      updateProfile,
      infoLoaded,
      setInfoLoaded,
    }}
  >
    {children}
  </UserContext.Provider>
);

export { UserProvider };
