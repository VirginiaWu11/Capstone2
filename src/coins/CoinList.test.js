import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router";
import CoinList from "./CoinList";
import { UserProvider } from "../testUtils";

it("matches snapshot", function () {
  const { asFragment } = render(
    <MemoryRouter>
      <UserProvider>
        <CoinList />
      </UserProvider>
    </MemoryRouter>
  );
  expect(asFragment()).toMatchSnapshot();
});
