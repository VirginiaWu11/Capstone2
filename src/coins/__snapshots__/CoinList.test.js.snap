// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="MuiContainer-root MuiContainer-maxWidthXl css-19r6kue-MuiContainer-root"
    >
      <div
        class="MuiBox-root css-gg4vpm"
      >
        <div
          class="MuiToggleButtonGroup-root css-1deg3qg-MuiToggleButtonGroup-root"
          role="group"
        >
          <button
            aria-label="list"
            aria-pressed="false"
            class="MuiButtonBase-root MuiToggleButton-root MuiToggleButton-sizeMedium MuiToggleButton-standard MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal css-ueukts-MuiButtonBase-root-MuiToggleButton-root"
            tabindex="0"
            type="button"
            value="list"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="ViewListIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M3 14h4v-4H3v4zm0 5h4v-4H3v4zM3 9h4V5H3v4zm5 5h13v-4H8v4zm0 5h13v-4H8v4zM8 5v4h13V5H8z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <button
            aria-label="module"
            aria-pressed="true"
            class="MuiButtonBase-root MuiToggleButton-root Mui-selected MuiToggleButton-sizeMedium MuiToggleButton-standard MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal css-ueukts-MuiButtonBase-root-MuiToggleButton-root"
            tabindex="0"
            type="button"
            value="module"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="ViewModuleIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M14.67 5v6.5H9.33V5h5.34zm1 6.5H21V5h-5.33v6.5zm-1 7.5v-6.5H9.33V19h5.34zm1-6.5V19H21v-6.5h-5.33zm-7.34 0H3V19h5.33v-6.5zm0-1V5H3v6.5h5.33z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
        <div
          class="MuiToggleButtonGroup-root css-1vuf753-MuiToggleButtonGroup-root"
          role="group"
        >
          <button
            aria-label="all coins"
            aria-pressed="true"
            class="MuiButtonBase-root MuiToggleButton-root Mui-selected MuiToggleButton-sizeMedium MuiToggleButton-standard MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal css-ueukts-MuiButtonBase-root-MuiToggleButton-root"
            tabindex="0"
            type="button"
            value="allcoins"
          >
            All Coins
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <button
            aria-label="watchllist"
            aria-pressed="false"
            class="MuiButtonBase-root MuiToggleButton-root MuiToggleButton-sizeMedium MuiToggleButton-standard MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal css-ueukts-MuiButtonBase-root-MuiToggleButton-root"
            tabindex="0"
            type="button"
            value="watchlist"
          >
            Watchlist
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3 css-1equabv-MuiGrid-root"
      >
        <p>
          No coins found
        </p>
      </div>
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1 css-ng6zl4-MuiGrid-root"
      >
        <div
          class="MuiBox-root css-125k91q"
        >
          <div
            class="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root"
          >
            <label
              class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary MuiFormLabel-filled css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="true"
              id="items-per-page"
            >
              Items Per Page
            </label>
            <div
              class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <div
                aria-expanded="false"
                aria-haspopup="listbox"
                aria-labelledby="items-per-page items-per-page"
                class="MuiSelect-select MuiSelect-outlined MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputSizeSmall css-jedpe8-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
                id="items-per-page"
                role="button"
                tabindex="0"
              >
                Sixteen
              </div>
              <input
                aria-hidden="true"
                class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
                tabindex="-1"
                value="16"
              />
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
                data-testid="ArrowDropDownIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7 10l5 5 5-5z"
                />
              </svg>
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-1z7n62"
                >
                  <span>
                    items per page
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
        <div
          class="css-1p5q5e5-MuiStack-root"
        >
          <nav
            aria-label="pagination navigation"
            class="MuiPagination-root MuiPagination-outlined css-1oj2twp-MuiPagination-root"
          >
            <ul
              class="MuiPagination-ul css-wjh20t-MuiPagination-ul"
            >
              <li>
                <button
                  aria-label="Go to previous page"
                  class="MuiButtonBase-root Mui-disabled MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-outlined MuiPaginationItem-circular MuiPaginationItem-outlinedPrimary Mui-disabled MuiPaginationItem-previousNext css-1y7coo4-MuiButtonBase-root-MuiPaginationItem-root"
                  disabled=""
                  tabindex="-1"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiPaginationItem-icon css-g2z002-MuiSvgIcon-root-MuiPaginationItem-icon"
                    data-testid="NavigateBeforeIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
                    />
                  </svg>
                </button>
              </li>
              <li>
                <button
                  aria-current="true"
                  aria-label="page 1"
                  class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-outlined MuiPaginationItem-circular MuiPaginationItem-outlinedPrimary Mui-selected MuiPaginationItem-page css-1y7coo4-MuiButtonBase-root-MuiPaginationItem-root"
                  tabindex="0"
                  type="button"
                >
                  1
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </button>
              </li>
              <li>
                <button
                  aria-label="Go to page 2"
                  class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-outlined MuiPaginationItem-circular MuiPaginationItem-outlinedPrimary MuiPaginationItem-page css-1y7coo4-MuiButtonBase-root-MuiPaginationItem-root"
                  tabindex="0"
                  type="button"
                >
                  2
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </button>
              </li>
              <li>
                <button
                  aria-label="Go to page 3"
                  class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-outlined MuiPaginationItem-circular MuiPaginationItem-outlinedPrimary MuiPaginationItem-page css-1y7coo4-MuiButtonBase-root-MuiPaginationItem-root"
                  tabindex="0"
                  type="button"
                >
                  3
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </button>
              </li>
              <li>
                <button
                  aria-label="Go to page 4"
                  class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-outlined MuiPaginationItem-circular MuiPaginationItem-outlinedPrimary MuiPaginationItem-page css-1y7coo4-MuiButtonBase-root-MuiPaginationItem-root"
                  tabindex="0"
                  type="button"
                >
                  4
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </button>
              </li>
              <li>
                <button
                  aria-label="Go to page 5"
                  class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-outlined MuiPaginationItem-circular MuiPaginationItem-outlinedPrimary MuiPaginationItem-page css-1y7coo4-MuiButtonBase-root-MuiPaginationItem-root"
                  tabindex="0"
                  type="button"
                >
                  5
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </button>
              </li>
              <li>
                <div
                  class="MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-outlined MuiPaginationItem-circular MuiPaginationItem-outlinedPrimary MuiPaginationItem-ellipsis css-1v2lvtn-MuiPaginationItem-root"
                >
                  …
                </div>
              </li>
              <li>
                <button
                  aria-label="Go to page 750"
                  class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-outlined MuiPaginationItem-circular MuiPaginationItem-outlinedPrimary MuiPaginationItem-page css-1y7coo4-MuiButtonBase-root-MuiPaginationItem-root"
                  tabindex="0"
                  type="button"
                >
                  750
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </button>
              </li>
              <li>
                <button
                  aria-label="Go to next page"
                  class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-outlined MuiPaginationItem-circular MuiPaginationItem-outlinedPrimary MuiPaginationItem-previousNext css-1y7coo4-MuiButtonBase-root-MuiPaginationItem-root"
                  tabindex="0"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiPaginationItem-icon css-g2z002-MuiSvgIcon-root-MuiPaginationItem-icon"
                    data-testid="NavigateNextIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
                    />
                  </svg>
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </button>
              </li>
            </ul>
          </nav>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
